### ISSUE TITLE
Getting "Cannot read properties of undefined (reading 'map')"

### ISSUE THREAD

**Issue by nvadluri** [loblaw-sre/backstage-plugin-gitlab#38]
<img width="1534" alt="Screen Shot 2021-11-10 at 12 24 36 PM" src="https://user-images.githubusercontent.com/84935639/141163276-d2ccd1ed-7c95-40df-81b5-06e26b8fcd22.png">

mine is a standalone app and integrated this plugin as in readme file instructions including, app-config updates with GitLab integration and proxy details. I am still getting this error. (I masked my hostname)

**Comment by nvadluri** [comment]
when I click on the GitLab tab, backpage is trying to load the GitLab build jobs information for a second, then immediately, getting the error.
![IMG_1372](https://user-images.githubusercontent.com/84935639/141209432-940ef9f4-2d4e-41f2-a9b1-2b53ecf8685f.jpg)
..

**Comment by nvadluri** [comment]
![IMG_1372 2](https://user-images.githubusercontent.com/84935639/141209854-142675c5-5193-49e0-9da5-b97e113cd401.jpg)
for 1 second, when I clicked GitLab tab, it is trying to load, then immediately, getting the error like mentioned...

**Comment by nvadluri** [comment]
Can someone please help me fixing the issue? I tried in many ways, but getting the same error all the time. Thank you in advance.

**Comment by warmfusion** [comment]
I had this issue, it seemed to be something related to permissions.

Check your backstage-api logs for requests to the proxy/gitlabci endpoint and see if you open localhost:7000 on that path if you get anything.

I managed to fix this by ensuring my gitlab credentials had access to the projects

**Comment by nvadluri** [comment]
I tested to see if the backstage can pull the config files from the Gitlab repo, and found it is pulling config files from gitlab to add existing components, it is pulling successfully, and you see the image(2nd image), it is trying to load the GitLab jobs, then it's giving me the error.
These are the below log chunks giving 200 

![image](https://user-images.githubusercontent.com/84935639/141373008-5b10eab4-b1f6-4eed-a5ab-a89e692c1d04.png)


<img width="1537" alt="Screen Shot 2021-11-11 at 4 34 31 PM" src="https://user-images.githubusercontent.com/84935639/141373076-b920b32f-7200-4915-bf4e-9c63dbcf13ca.png">

**Comment by nvadluri** [comment]
@warmfusion when I curl this from the backstage pod I am getting 
curl --header "PRIVATE-TOKEN: mytoken" "https://gitlab.com/api/proxy/gitlabci/projects/28586484/merge_requests?per_page=20"
<html><body>You are being <a href="https://gitlab.com/users/sign_in">redirected</a>.</body></html>

or 
curl --header "PRIVATE-TOKEN: mytoken" "https://gitlab.com/api/v4/personal_access_tokens?user_id=:9020478"
{"message":"401 Unauthorized"}

btw, when I do this 

curl -vk --header "PRIVATE-TOKEN: mytoken" "https://gitlab.com/api/proxy/gitlabci/projects/123456?"
*   Trying 172.65.**.**...
* TCP_NODELAY set
* Connected to gitlab.com (172.65.**.**) port 443 (#0)
* ALPN, offering h2
......
REMOVED CONTENT
......
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 256)!
< HTTP/2 302
< date: Fri, 12 Nov 2021 07:54:11 GMT
......
REMOVED CONTENT
......

<
* Connection #0 to host gitlab.com left intact
<html><body>You are being <a href="https://gitlab.com/users/sign_in">redirected</a>.</body></html>* Closing connection 0
 

To create the personal access token, I went here https://gitlab.com/-/profile/personal_access_tokens and created one with read_api access. Anything I am missing, can you please guide me if I am missing anything?

**Comment by warmfusion** [comment]
Check the [gitlab api docs](https://docs.gitlab.com/ee/api/#personalproject-access-tokens) for using their API

You're hitting `proxy/gitlabci/` which is a Backstage endpoint which basically acts as an auth proxy to gitlab using the tokens you configured.

- If you want to hit gitlab use; https://gitlab.com/api/v4/projects/28.... with PRIVATE-TOKEN
- If you want to use backstage proxy use; http://localhost:7000/api/proxy/gitlabci/projects/28.... *without* PRIVATE_TOKEN

**Comment by nvadluri** [comment]
I used now like below in both ways, still getting the same error (GITLAB_URL=http://localhost:7000)
proxy:
  '/gitlabci':
    target: '${GITLAB_URL}/api/proxy'
    allowedMethods: ['GET']
    headers:
      PRIVATE-TOKEN: '${GITLAB_TOKEN}'
      
      and also without token as you mentioned
      
      proxy:
  '/gitlabci':
    target: '${GITLAB_URL}/api/proxy'
    allowedMethods: ['GET']

**Comment by nvadluri** [comment]
@warmfusion 
 I used now like below in both ways, still getting the same error (GITLAB_URL=http://localhost:7000)
proxy:
  '/gitlabci':
    target: '${GITLAB_URL}/api/proxy'
    allowedMethods: ['GET']
    headers:
      PRIVATE-TOKEN: '${GITLAB_TOKEN}'
      
 and also without token as you mentioned
      
  proxy:
  '/gitlabci':
    target: '${GITLAB_URL}/api/proxy'
    allowedMethods: ['GET']

**Comment by satrox28** [comment]
Hi @nvadluri 

Thanks for trying out the Gitlab Plugin. 

Are you using the latest version 0.1.16 of Gitlab Plugin?.

Can you bump the backstage packages from the root dir
```
#yarn backstage-cli versions:bump
```
4. Add proxy config:
```
  '/gitlabci':
    target: '${GITLAB_URL}/api/v4'
    allowedMethods: ['GET']
    headers:
      PRIVATE-TOKEN: '${GITLAB_TOKEN}'

```
Default GitLab URL: https://gitlab.com
GitLab Token should be with of scope read_api and can be generated from this URL

Have you replaced the Gitlab URL and TOKEN with the actual values. The token which is created should be of scope read_api.

Please check all above and let us know.

**Comment by nvadluri** [comment]
Thank you for the reply. I tried as you said, but still getting the same error.  Can you please help me solving this issue?

1. Are you using the latest version 0.1.16 of Gitlab Plugin?.
Yes 
    "@loblaw/backstage-plugin-gitlab": "^0.1.16",
    
  
2. Can you bump the backstage packages from the root dir
 Yes, I did just now
 
 3. Add proxy config:
 
 yes, I added and below is my app-config.yaml file chunk, and also I replased with actula values and the api is with read_api
   gitlab:
    - host: gitlab.com
      token: ${GITLAB_TOKEN}

proxy:
  '/gitlabci':
    target: '${GITLAB_URL}/api/v4'
    allowedMethods: ['GET']
    headers:
      PRIVATE-TOKEN: '${GITLAB_TOKEN}'
      
  4. Add catalog-info.yaml with below changes ?
  
  I tried in both ways, still getting the same error, but right now, my config is with project-slug as below (I am pulling this catelog-info.yaml file from gitlab itself)

  apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: gitlab-cicd
  # This is an extra long description
  description: ci/cd for streaming app deployment
  annotations:
    gitlab.com/project-slug: *****/global-digital-tech/commerce-and-transactions/microservices-template-app/baseapp
spec:
  type: service
  lifecycle: Production
  owner: post-purchase
  system: create-order

**Comment by satrox28** [comment]
Hi @nvadluri 

Can you try with project-id annotation.

**Comment by nvadluri** [comment]
I tried that too, got the same error, then only I moved to project-slug.

**Comment by nvadluri** [comment]
@satrox28 I tried that too, got the same error, then only I moved to project-slug.

**Comment by nvadluri** [comment]
@satrox28 any other suggestions ?

**Comment by satrox28** [comment]
Hi @nvadluri 

Do you see response from the api end point?. 

http://localhost:7000/api/proxy/gitlabci/projects/<project_id>

**Comment by nvadluri** [comment]
yes, I can see the response. 

"root@backstage-65d7cb795c-t5n8h:/app# apt update & apt upgrade^C
root@backstage-65d7cb795c-t5n8h:/app# curl --header "MYToken" http://localhost:7000/api/proxy/gitlabci/projects/2858***
{"id":2858****,"description":null,"name":"baseapp","name_with_namespace":"***** / Global Digital Technology / Commerce And Transactions / Microservices Template App / baseapp","path":"baseapp","path_with_namespace":"****/global-digital-tech/commerce-and-transactions/microservices-template-app/baseapp","created_at":"2021-08-03T15:50:48.091Z","default_branch":"main","tag_list":[],"topics":[],"ssh_url_to_repo":"git@gitlab.com:****/global-
....
....
....

**Comment by satrox28** [comment]
Hi,

Do you see any console log on the browser.

**Comment by nvadluri** [comment]
<img width="1042" alt="Screen Shot 2021-11-15 at 1 06 28 PM" src="https://user-images.githubusercontent.com/84935639/141832435-62ae03dc-0a56-43fd-b65a-affdb7faffd9.png">

**Comment by satrox28** [comment]
Hi @nvadluri 

Have you tried restarting your application. I feel the catalog is not refreshed.

Can you try with below annotation in your catalog-info.yaml.

gitlab.com/project-id: '1794617'

**Comment by nvadluri** [comment]
<img width="1499" alt="Screen Shot 2021-11-16 at 1 00 07 PM" src="https://user-images.githubusercontent.com/84935639/142040439-33ce4b2a-efd6-4a33-a98f-fe9f8f61cbe1.png">

Wow! I can load without any issues which means the issue with my access to the API key to plans? Please suggest the solution.

**Comment by satrox28** [comment]
Hi @nvadluri

Can you able to try this from your local for the project id which you are having issue.

curl -X GET -H "PRIVATE-TOKEN: <ACCESS_TOKEN>" "https://gitlab.com/api/v4/projects/<project_id>"

**Comment by nvadluri** [comment]
@satrox28 I am getting a response from my local machine

nvadluri@C02D72B7MD6T ~ % curl -X GET -H "PRIVATE-TOKEN: Mytoken" "https://gitlab.com/api/v4/projects/285*****"
{"id":285*****,"description":null,"name":"XXXXX","name_with_namespace":"XXXXX / Global Digital Technology / Commerce And Transactions / Microservices Template App / XXXXX","path":"XXXXX","path_with_namespace":"XXXXXX/global-digital-tech/commerce-and-transactions/microservices-template-app/XXXXX","created_at":"2021-08-03T15:50:48.091Z","default_branch":"main","tag_list":[],"topics":[],"ssh_url_to_repo":"

**Comment by satrox28** [comment]
@nvadluri Can you confirm if the backstage application using the same access token?.

If you still face issue, please right click on the Gitlab page and choose inspect-->console and refresh the page to check the error messages. https://developer.chrome.com/docs/devtools/open/. Can you able to perform this operation?.

**Comment by nvadluri** [comment]
@satrox28 Yes, the same access token.

and below is the console error messages screenshot
<img width="954" alt="Screen Shot 2021-11-18 at 2 00 45 AM" src="https://user-images.githubusercontent.com/84935639/142368275-82cd6448-c410-48d1-9cb6-ca160a23c81c.png">

**Comment by satrox28** [comment]
Hi @nvadluri I can see the requests are getting 403, on the developer console can you click on Network tab and check a particular request and its response.

**Comment by nvadluri** [comment]
@satrox28  
<img width="558" alt="Screen Shot 2021-11-18 at 2 20 50 AM" src="https://user-images.githubusercontent.com/84935639/142370265-4d22cf2a-552e-4ffb-b54c-9c7aeb5098b3.png"
<img width="558" alt="Screen Shot 2021-11-18 at 2 20 22 AM" src="https://user-images.githubusercontent.com/84935639/142370305-7430186a-f585-4ae9-a82b-664493fecdc8.png">
>

**Comment by satrox28** [comment]
Hi @nvadluri 

Thanks for the screenshot. If you select languages, contributors, pipelines from the above list, can you see response?.

**Comment by nvadluri** [comment]
yes, I can see the details.
<img width="555" alt="Screen Shot 2021-11-18 at 10 08 58 AM" src="https://user-images.githubusercontent.com/84935639/142442334-55dad499-40bf-45f3-b6ad-5861ac869ae1.png">
<img width="555" alt="Screen Shot 2021-11-18 at 10 08 22 AM" src="https://user-images.githubusercontent.com/84935639/142442357-7467bdb8-67aa-42aa-aa4e-8c6e804e81c3.png">

**Comment by nvadluri** [comment]
Yes, I can see the response.
<img width="555" alt="Screen Shot 2021-11-18 at 10 08 58 AM" src="https://user-images.githubusercontent.com/84935639/142442911-d0da575b-f886-4100-b07c-530221782db6.png">
<img width="555" alt="Screen Shot 2021-11-18 at 10 08 22 AM" src="https://user-images.githubusercontent.com/84935639/142442939-26c85996-e5f0-4dd3-858e-556c3a07ebed.png">

**Comment by nvadluri** [comment]
@satrox28 I think I found the reason..The project I am trying to pull the information is actually mirrorred from github (the main code base is in github, and we mirrored the repo into gitlab). for the testing purpose, I created a dummy project directly on gitlab, and tried to pull the information on backstage, it worked. Below is the screen shot of the dummy project. What is the solution for such projects(mirroring from gitlab) ?

<img width="1541" alt="Screen Shot 2021-11-18 at 10 58 02 PM" src="https://user-images.githubusercontent.com/84935639/142568017-7c4d564a-8506-4b3f-8ba2-93fac2671ba6.png">

**Comment by satrox28** [comment]
Hi @nvadluri 

When we mirror a repository it doesn't copy the MR from the source. Since you are not seeing any MR on Gitlab, getting the map error, we will handle this exception in the future release.

**Comment by drobakowski** [comment]
Hi @nvadluri, I'm not sure if it helps you, but for me removing the `''` from PRIVATE-TOKEN did the trick:

```yaml
proxy:
  '/gitlabci':
    target: 'https://.../api/v4'
    allowedMethods: ['GET']
    headers:
      PRIVATE-TOKEN: ${GITLAB_TOKEN}
```
instead of:
```yaml
    headers:
      PRIVATE-TOKEN: '${GITLAB_TOKEN}'
```
Works for me.

**Comment by nvadluri** [comment]
@satrox28 I created one dummy project directly on GitLab, and I can see MR for this repo, it is on the above image which I uploaded.

@drobakowski thank you for the reply...the issue has been fixed but the only thing is all repos we have right now are mirrored from GitHub to GitLab...that is the issue...may be in the future we will be good, as we are in a plan to move completely to GitLab.

**Comment by satrox28** [comment]
Hi @nvadluri 

I was asking if the mirrored repo shows MR on Gitlab, if you have PR on Github. Mirroring will not copy MR's.

**Comment by nvadluri** [comment]
@satrox28  Sorry, maybe I am confused. Can you please brief it a little more? Right now, I can only see MRs for the Repo which is not mirrored. for Mirrored repos, I am getting map errors.

**Comment by satrox28** [comment]
Hi @nvadluri 

The repo for which you are getting map error, does it show Merge Request on Gitlab(mirrored repo) and the actual Github repo, does it have any Pull Request.

**Comment by ryharrin** [comment]
I've run into this issue as well. All of my repos are mirrored from Github (where they almost all have PRs) and none of them have any merge requests in Gitlab. 

The project object returned shows `"merge_requests_enabled": false`, which must be the default for mirrored projects.  I get a 403 when trying to get merge requests.

After I enabled merge requests on a project, the plugin loads properly and the GitLab api correctly returns an empty array for merge requests.

**Comment by jwnmulder** [comment]
Same here, CI/CD option was disabled which resulted in a 403 for the `pipelines` url followed by the `Cannot read properties of undefined (reading 'map')` error. After enabling CI/CD for that specific project the page is loading fine.

**Comment by satrox28** [comment]
We will check and work on an appropriate fix when the merge_request/CICD disabled on the Project settings @ryharrin @jwnmulder @nvadluri. Please feel free to contribute, if interested.

**Comment by nvadluri** [comment]
@jwnmulder yes, after enabling merge requests on the GitLab project pipeline's settings, I could load the project details on the backstage even it is mirrored from GitHub.

<img width="721" alt="Screen Shot 2021-12-22 at 3 12 58 PM" src="https://user-images.githubusercontent.com/84935639/147150266-3c18cba6-b5bb-4043-bcce-2d340050f373.png">

@satrox28 yes, having that feature may resolve these confusions.

**Comment by ryharrin** [comment]
@satrox28 - I just created a PR that I believe should fix this, but I was having trouble testing it locally. 

https://github.com/loblaw-sre/backstage-plugin-gitlab/pull/60

**Comment by satrox28** [comment]
Hi @ryharrin @nvadluri @jwnmulder 

It's fixed on v0.1.19

If Merge Request disabled on a project, users won't be seeing any map dependency error.
